draft-ietf-ppm-dap.md

```diff
@@ -1722,11 +1738,16 @@ preparation continuation messages constructed in the previous step. The
 The Leader MUST authenticate its requests to the Helper using a scheme that
 meets the requirements in {{request-authentication}}.
 
-The Helper's response will be an `AggregationJobResp` message (see
-{{aggregation-helper-init}}). The response's `prepare_resps` must include
-exactly the same report IDs in the same order as the Leader's
-`AggregationJobContinueReq`. Otherwise, the Leader MUST abort the aggregation
-job.
+After receiving HTTP status 202 Accepted, the Leader proceeds to poll the
+aggregation job by sending requests to
+`GET {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` until
```

Comment:
The same URI is used for each prep step. Do we foresee this being an issue? Perhaps this is actually desirable from an idempotency standpoint?

cc/ @tgeoghegan, @branlwyd

Reply:
It was already the case that the same URI was used regardless of which step we were on. Asynchronous handling of aggregation jobs shouldn't change that.

Reply:
Since this is async interaction, leader can send two POST requests to this URI, one with valid and another with invalid body. When we query for status with GET, it would be ambiguous to interpret the response. Two options to resolve this -
* Maintain a requestID so a GET can be pointed to a specific requestID
* Somehow early reject 2nd request instead of always returning 202

Reply:
Invalid in what sense? If the message is malformed or something, then the Helper would abort the request. I think it's the Helper's job to ensure that one and only one request is accepted.

Reply:
When two requests are accepted with 202 on the same URL, there is no way to know the response fetched later belongs to request1 or request2. Hope that clarifies.

Reply:
When would this happen? Are you thinking of something like a race condition, where the Leader erroneously sends two POSTs at the same time for the same aggregation job?

Reply:
It could happen for various reasons -
1. Leader did not receive response due to network error and retries.
2. Leader incorrectly tracks the step count and sends continue request with invalid step count.
3. Two leader instances sends the same request
4. Other malicious reasons

I think without request to response correlation, it adds stateful processing responsibility on leader and still does not give a way to track this correlation in a consistent way.

Reply:
Given that the relevant VDAF functions are deterministic, when both aggregators are honestly following the protocol, there is exactly one correct next request for a given aggregation job[1].

Therefore, I suspect the semantics we want are that the _first_ request received for a given aggregation request determines the correct request; if the Leader sends multiple, non-identical requests for the same aggregation job, something has gone wrong with the Leader & failure is OK. That is, we want to detect repeated non-identical requests, but the proper way to respond to them is with failure rather than tracking both requests.

[1] This is true semantically but may not be true up to encoding; though DAP messages are deterministically-encoded if the underlying VDAF messages are, and to my knowledge all current VDAFs use messages which are deterministically-encoded.

Reply:
VDAF preparation is deterministic. Given the current _strict_ ordering of reports in the job, there should be exactly one valid next message.

Reply:
Well, one thing that can change is that the Helper's decryption key gets rotated out, causing the report to be rejected.

Reply:
I guess I don't get what response Helper gives in these cases. Response to last request or response to first request?

If it is giving response to first request to all - then an incorrect submission (request2) gets a valid response.
If it is giving response to last request to all - then a valid submission receives error response without knowing why.

Besides this, My understanding is, leader does not have to be good citizen and can learn things and hence Helper implements
* Prevention of replay attacks
* Prevention of same share belonging to several batches etc.
Is that right understanding and still holds good here?

Reply:
Yes, both the Leader and Helper are responsible for preventing replays, including across batches.

Reply:
Okay. I still need to understand this part -

> 
> I guess I don't get what response Helper gives in these cases. Response to last request or response to first request?
> 
> If it is giving response to first request to all - then an incorrect submission (request2) gets a valid response.
> If it is giving response to last request to all - then a valid submission receives error response without knowing why.

Reply:
>     1. Leader did not receive response due to network error and retries.

If the body of each request is the same, then so should be the Helper's respond. Otherwise, the Leader is in error, and it's up to the Helper to decide which order they're processed in.

>     2. Leader incorrectly tracks the step count and sends continue request with invalid step count.

The Leader is in error. The Helper will abort, but it [might be possible to retry](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-4.5.2.3).

>     3. Two leader instances sends the same request

Same as for 1 I think.

>     4. Other malicious reasons

Even if the Leader is malicious, then it's still up to the Helper to enforce invariants like anti-replay. However there are probably any number of ways the Leader can break correctness.

Reply:
If two mutating requests results in two different responses, then it is not deterministic to let Helper decide what to do with them. I think it is deterministic if the helper respond with the unique URL to track the POST request status. I think this behavior should not depend on what Leader's behavior is.

Note that, this would not be a problem if the response is synchronous since it comes with request-response correlation built-in.

Reply:
Let's maybe take a step back for a second: what property do you want DAP to have that this PR doesn't have? It sounds like you want some way for the Leader to recover if it makes a mistake.

It sounds like you're suggesting the following mechanism: the Leader chooses a unique identifier for each request (probably at random?), which the Helper is supposed to echo back in its response. Is this what you're suggesting?

If so, what errors does this help the Leader recover from? What errors does it not help it recover from?

Reply:
> When two requests are accepted with 202 on the same URL, there is no way to know the response fetched later belongs to request1 or request2.

Concretely, in Janus, each insertion of aggregation job is transactional since we use a relational database. The operations are serialized--a second request will reject with a database conflict error. That is, the only scenario where we would return 202 for a repeated request is if the second request is exactly identical to the first.

However, I don't think we have any requirement for transactionality, either in this PR or generally in DAP. An aggregator that lacks transactionality _could_ accept multiple PUTs, and indeed it would be implementation-defined as to which request the subsequent GET request is responding to.

So I think @suman-ganta is correct. Perhaps we should add clarification that the helper should only accept one PUT request?

Reply:
Eh. Reading through it again as I try to write it:
```
Changing an aggregation job's parameters is illegal, so further requests to
`PUT /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` for the same
`aggregation-job-id` but with a different `AggregationJobInitReq` in the body
MUST fail with an HTTP client error status code.
```
I feel we are adequately signaling that only one PUT request should be accepted. i.e. a helper that is accepting multiple different PUT requests is not following the spec. Similarly, I think the continue section with step recovery strongly implies the need for transactionality/consistency.

So I think my point earlier:
> However, I don't think we have any requirement for transactionality, either in this PR or generally in DAP.

is not valid. Sorry, I should have given a close read before commenting.

So @suman-ganta, I think the questions in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/564#discussion_r1679566081 still apply.

---

> Also @suman-ganta and @kristineguo point out offline that, if a POST comes in for step 2 while we're processing step 1, then we should abort the POST request.

I think stepMismatch covers this? If the helper receives a request for step 2 while step 1 is still in flight, the aggregation job will still be on step 1, thus being errored out per this paragraph.
```
Next, the Helper checks if the continuation step indicated by the request is
correct. (For the first `AggregationJobContinueReq` the value should be `1`;
for the second the value should be `2`; and so on.) If the Leader is one step
behind (e.g., the Leader has resent the previous HTTP request), then the Helper
MAY attempt to recover by sending HTTP 202 status Accepted and discontinuing
any further work. In this case it SHOULD verify that the contents of the
`AggregationJobContinueReq` are identical to the previous message (see
{{aggregation-step-skew-recovery}}). Otherwise, if the step is incorrect, the
Helper MUST abort with error `stepMismatch`.
```

Reply:
I don't know how often this scenario would actually come up in reality, but imagine a scenario where the Leader sends out two synchronous requests when recovering:
- `POST /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` for step 1 (already computed)
  - Helper may return `202 Accepted` here if the request matches the previous message.
- `POST /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` for step 2 (new)

Does the Helper return `202 Accepted` for that second POST request? If the Helper does, what should the Helper return when the Leader does `GET /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}`? The response for step 1 or 2?

Reply:
To me, the confusion can arise because we're using the same request path for multiple different requests (one for each step). We could potentially alleviate this by making the request path more explicit by adding the step into the requestPath, so that the Helper always knows which step the Leader is sending/requesting.

Reply:
Good scenario, I think it's plausible. Perhaps not under normal operations, but it is a series of operations that I think would be accepted. Let me think about it out loud:

I'm going to assume that aggregation jobs processing is perfectly transactional, and that these simultaneous requests don't leave the job in some indeterminate state. I believe these are valid assumptions, but stop me if they're not.

---
Consider the first scenario, where the order of requests is serialized to request 1 (i.e. your bullet point 1), then request 2.

Request 1: POST returns 202 Accepted, any subsequent GET requests _would_ return 200 with the proper body, since it was already computed.
Request 2: POST returns 202 Accepted, any subsequent GET requests would return 202 until the aggregation job is complete.

I don't think anything is surprising here.

---
Second scenario, the order of requests is flipped.

Request 2: POST returns 202 accepted, any subsequent GET requests would return 202 until the aggregation job is complete.
Request 1: Step recovery kicks in, because we're trying to POST for step 1. Uh oh. I think you're right that this situation is ambiguous. For step recovery to actually be possible, we'd need to abort continuation to step 2. Or the leader would need to wait for step 2 to be done, then try step recovery again.

---

> To me, the confusion can arise because we're using the same request path for multiple different requests (one for each step). We could potentially alleviate this by making the request path more explicit by adding the step into the request path, so that the Helper always explicitly knows which step the Leader is requesting a response for.

I had originally considered having different paths for different steps, but couldn't find a scenario where this was strictly necessary, but I think you've identified one.

Let me dwell on it for a bit, anyone else have thoughts as well?

Reply:
@inahga I don't think that we should be prescribing serializable transactions here - this is very limiting and operationally challenging for partitioned helpers.

With our helper, we've found that this aspect of the replay checks are the hardest part of the system to scale up - they eventually become a breaking point for the system unless you shard, and that has availability implications under draft 11 due to the synchronous nature of the protocol.

Ideally we'd make things idempotent, or defer when transactional things have to occur until after jobs have been accepted for processing.

Reply:
Fair points. It sounds like then simultaneous POSTs with different steps would be exceptionally painful for you, because it'll become indeterminate which request is being responded to for the subsequent GET. Does that sound right?

Reply:
An alternative to step specific path is to return 202 with a trackable URL. This URL contains unique ID in it generated by the helper that leader can GET. It responds with status of that specific submission.

Another simpler way is to synchronously check if there is a pending `continue` submission to be processed for the job, if so, abort synchronously with an error.

Reply:
Those resolutions all seem plausible to me.

OK to summarize the thread direction: sharing the same URI for each step introduces ambiguity as to which request takes priority when requests for different steps come in simultaneously.

The solutions to this are:
1. Per step request routing, e.g. `{task-id}/aggregation_jobs/{aggregation-job-id}/{step-number}`.
1. Provide trackable URLs per PUT/POST, such that subsequent GETs can use it to get the proper response.
1. Add prerequisite that aggregators MUST abort if another init/continue is still in progress.
1. Do nothing--the scenario we painted above is admittedly somewhat pathological.

I favor solution one, because I think it'll simplify step recovery substantially, because the leader can straightforwardly GET step it has lost track of.

Does anyone else who hasn't chimed in in a while have thoughts, @branlwyd @cjpatton?

Reply:
You've got it!

Reply:
Thanks @inahga for the super succinct summary!

With the major caveat that I've not implemented the Leader: I'm inclined to take 4. If I understand correctly, this case occurs because something has gone seriously wrong with the Leader. Generally speaking, [we have stricter operational requirements for the Leader than Helper](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-5.1.2-1). Thus I think we could reasonably "give up" and accept the data loss in this case.

That said, those that have implemented the Leader get a bigger vote share. I would nudge us towards 1 as it seems like the simplest way to get what we need. We might tweak it a bit to maintain the "resource API" semantics: `{task-id}/aggregation_jobs/{aggregation-job-id}?step={step-number}`

Reply:
It also seems like this will be helpful for https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/569.

Reply:
I like #3 or #1.

#3 feels the cleanest since, in normal operation, only a single aggregation step will be computed at once; it's OK to fail if multiple steps are attempted at once. The step would be considered "complete" (allowing another step to be computed) once the Helper considers its response to be retrieved at least once.

I'm not sure if #3 would step on the toes of some eventual-consistency scheme, however. If so, #1 seems the next cleanest.

Reply:
I also agree with @cjpatton that for this scenario to happen the leader must be in some weird state, but I have also not implemented a leader. My thinking is that the leader would never even send a request for step 2 before having successfully polled the response for step 1 to completion, meaning the race condition could never occur. And as a helper, when I receive a request for step 2 after being done with step 1 I would just start aggregating step 2 and discard step 1 data. Meaning it would not be ambiguous for the helper which step the `GET` would be related to.

Now, despite this confusion on my part, I would still vote for a solution which includes the step number in the request (number 1), since it probably makes debugging/observability easier because it makes the intent of the leader obvious

Reply:
One curiosity about solution 1 is that it makes the step number in the `AggregationJobContinueReq` somewhat redundant. We may be tempted to take a wire breaking change to prune that field, as I don't think we would want to keep vestigial fields around.

---

> I'm not sure if 3 would step on the toes of some eventual-consistency scheme, however.

I had assumed so, hence my vote for 1. Each replica of an aggregator needs to have a perfectly consistent view of aggregation job status to be able to determine whether an aggregation job is in progress.

In some architectures, I believe this is really, really tricky. Naively I may implement async jobs as POST/PUTs pushing onto a message queue. Some dedicated message queueing software do not support introspecting the queue, e.g. [RabbitMQ](https://lists.rabbitmq.com/pipermail/rabbitmq-discuss/2011-August/014270.html). Some do, e.g. Sidekiq, but I believe that's because it's backed by a RDBMS.

So by imposing the requirement of 3, we have limited the implementation selection to software that can introspect the queue. Which, without having looked into it much, may only be well supported by relational datastores. _Even then_, depending on the sharding scheme being used, would a single request require introspecting the queues of all the shards?
---

I agree with all folks that the scenario only arises in the presence of a badly malfunctioning leader. But I think the fact that both requests can return 2xx responses as written in the spec is problematic, since only one request is obviously correct (i.e. the move to step 2), so I'm not favoring Do Nothing atm.

Reply:
If you don't want to make a wire change, we could do:
`PUT {task-id}/aggregation_jobs/{aggregation-job-id}`
`GET {task-id}/aggregation_jobs/{aggregation-job-id}?step={step-number}`

The biggest issue is the `GET` has no way to explicitly identify the response it wants to retrieve. Leaving it up to the Helper to implicitly assume which step the Leader wants a response for leaves opportunity for (admittedly contrived) scenarios such as the one we've been discussing. But simply adding an explicit resource identifier of some sort, we avoid any potential for this confusion.

Hence why I'm partial to no.1, with step either as a query parameter or path component.

Reply:
@inahga it sounds like 1 is the option with the most consensus. Let's take it for this PR.

Reply:
> If you don't want to make a wire change, we could do:

I meant the definition of `AggregationContinueResp`. It's defined as.
```
struct {
  uint16 step;
  PrepareContinue prepare_continues<1..2^32-1>;
} AggregationJobContinueReq;
```

We may be tempted to remove the `step` field, since if `step` is a field of the URI, it's no longer necessary. That would be a simplification, but a wire breaking change.

Reply:
> If you don't want to make a wire change, we could do:

Sorry, I have Monday brain, I see what you're getting at now. We could let `AggregationJobContinueReq` retain the `step` field, retain calling `POST` as-is, but only use the `step` query parameter in the GET request. That would not be a wire breaking change.

Reply:
I've sketched out how this looks in the latest commit, editorial changes still probably required. I went with striking the `step` field, mostly for consistency (don't need the same data in two different locations), and debug-ability (easier to see what the leader is thinking about through HTTP status logs).

I've also moved from POST to PUT for continuation, since we use separate URIs for each step. This lets us use PUT semantics, i.e. idempotency.

That idempotency leads me to strike the entire job recovery section. Now I believe recovery is implicitly possible. If the Leader loses the acknowledgement of the PUT request, it can simply retransmit. Likewise with losing the response to the GET requests. Hopefully that's not overeager, as that seems like a pretty big win for simplicity of the text.

I've replaced it with a small paragraph that allows the helper to discard information about old steps, so that it doesn't have to retain the entire step history of a (theoretical) VDAF with many steps, but allows for the flexibility of recovering the whole step history (a-la https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/569). I anticipate implementing this as `if step < job.step - 1 { return Err }`.

Reply:
@inahga this is a big and somewhat contentious change, and for that reason, we need to make sure to implement the smallest change necessary. Please revert all changes except for adding the step parameter to the URL.

You are probably right about the simplifications we can make, but it will help everyone if we handle this in separate PRs. Feel free to just stick those in a draft PR for now.

Reply:
Makes sense, I have done so.

Reply:
I'm closing this thread! @kristineguo or @suman-ganta please chime in if there's more to address.